src/github/provider.py

```diff
@@ -0,0 +1,239 @@
+
+        Args:
+            owner (str): The owner of the repository.
+            repo_name (str): The name of the repository.
+            pr_number (int): The pull request number.
+            comment_type (CommentType): The type of comments to retrieve (issue or review).
+
+        Returns:
+            list[PRComment]: A list of PRComment objects representing the comments of the specified type.
+        """
+        if comment_type == CommentType.ISSUE_COMMENT:
+            comments = self.api.issues.list_comments(owner=owner, repo=repo_name, issue_number=pr_number)
+        else:
+            comments = self.api.pulls.list_review_comments(owner=owner, repo=repo_name, pull_number=pr_number)
+
+        return [
+            PRComment(
+                id=comment['id'],
+                type=comment_type,
+                user=comment['user']['login'],
+                user_type=comment['user']['type'],
+                diff_hunk=comment.get('diff_hunk'),
+                body=comment.get('body'),
+                created_at=comment['created_at'],
+                in_reply_to_id=comment.get('in_reply_to_id'),
+                html_url=comment['html_url']
+            )
+            for comment in comments
+        ]
+
+    def list_pr_comments(self, owner: str, repo_name: str, pr_number: int) -> list[PRComment]:
+        """
+        Lists all the comments associated with a GitHub pull request.
+
+        This method retrieves both issue comments and review comments for the specified pull request number. 
+        It returns a list of `PRComment` objects, which contain the relevant details for each comment, such as the user, diff hunk, body, and creation timestamp.
+
+        The comments are sorted by their creation timestamp in ascending order.
+
+        Args:
+            owner (str): The owner of the repository.
+            repo_name (str): The name of the repository.
+            pr_number (int): The number of the pull request to fetch comments for.
+
+        Returns:
+            list[PRComment]: A list of `PRComment` objects representing the comments for the specified pull request.
+        """
+        issue_comments = self._get_pr_comments(owner, repo_name, pr_number, CommentType.ISSUE_COMMENT)
+        review_comments = self._get_pr_comments(owner, repo_name, pr_number, CommentType.REVIEW_COMMENT)
+        return sorted(issue_comments + review_comments, key=lambda x: x.created_at)
+    
+    def get_patch(self, pr_file: dict) -> str:
+        """
+        Retrieve the patch (changes) for a file in a pull request.
+
+        Args:
+            pr_file (dict): The file information dictionary from the GitHub API.
+
+        Returns:
+            str: The patch content or 'no changes' if not available.
+        """
+        return pr_file.get('patch', 'no changes')
+
+    def _to_PullRequest(self, owner: str, repo_name: str, pr: dict, username: str) -> PullRequest:
```

Comment:
nit: I'd probably simplify the signature to only `(self, pr: dict)` since `owner`, `repo_name` and `username` can be retrieved from `pr`.

I have those changes in my local already so feel free to just merge and I'll add on

Reply:
Thanks for the help